### PR TITLE
Add Main chare

### PR DESCRIPTION
**Pull request by kidder**

### CODE REVIEW

File: src/Parallel/Main.hpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Exit.hpp"
+#include "Parallel/TypeTraits.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+#include "Parallel/Main.decl.h"
+
+template <typename... Elements, typename NiaryOp, typename... Args>
+constexpr void typelist_fold(const typelist<Elements...>& /*meta*/,
```
Comment (nilsdeppe) [inline src/Parallel/Main.hpp]: Do something with this

File: src/Parallel/Main.hpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+#include <charm++.h>
+#include <initializer_list>
+#include <type_traits>
+
+#include "Informer/Informer.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Exit.hpp"
+#include "Parallel/TypeTraits.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
```
Comment (nilsdeppe) [inline src/Parallel/Main.hpp:18]: TaggedTuple before TMPL? clang-format

File: src/Parallel/Main.hpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+namespace Parallel {
+
+/// \ingroup Parallel
+/// The main function of a Charm++ executable.
+///
+/// Metavariables must define the following:
+///
+/// \tparam Metavariables a struct containing metavariables
```
Comment (nilsdeppe) [inline src/Parallel/Main.hpp]: remove this

File: src/Parallel/Main.hpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  static_cast<void>(std::initializer_list<char>{
+      (static_cast<void>(op(Elements{}, args...)), '0')...});
+}
+
+namespace Parallel {
+
+/// \ingroup Parallel
+/// The main function of a Charm++ executable.
+///
+/// Metavariables must define the following:
```
Comment (nilsdeppe) [inline src/Parallel/Main.hpp:27]: finish this

File: src/Parallel/Main.hpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+///
+/// \tparam Metavariables a struct containing metavariables
+template <typename Metavariables>
+class Main : public CBase_Main<Metavariables> {
+ public:
+  using tentacle_list = typename Metavariables::tentacle_list;
+
+  explicit Main(CkArgMsg* msg);
```
Comment (nilsdeppe) [inline src/Parallel/Main.hpp]: noexcept

File: tests/Unit/Parallel/ParallelTestChares.hpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                                   const_global_cache_proxy);
+
+  explicit TestBoundArrayChare(CkMigrateMessage* /*msg*/) {}
+
+  int my_index() const noexcept { return this->thisIndex; }
+
+ private:
+  Parallel::CProxy_ConstGlobalCache<Metavariables> const_global_cache_proxy_;
+};
+
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/ParallelTestChares.hpp:87]: The `// template definitions` comment would be helpful

File: tests/Unit/Parallel/Test_Main.cpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  Parallel::printf("Calling setup_error_handling on processor %i\n",
+                   Parallel::my_proc());
+}
+
+// clang-tidy: may throw an exception tat cannot be caught [cert-err58-cpp]
+Catch::Session session;  // NOLINT
+
+struct TestMetavariables;
+
+std::unique_ptr<const Parallel::CProxy_ConstGlobalCache<TestMetavariables>>
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_Main.cpp]: raw pointer?

Comment (nilsdeppe) [reply tests/Unit/Parallel/Test_Main.cpp]: and a comment as to why this is a thing at all

File: tests/Unit/Parallel/Test_Main.cpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    int result = session.run();
+    if (0 != result) {
+      Parallel::abort("A catch test has failed");
+    }
+    return phase::Exit;
+  }
+};
+
+SPECTRE_TEST_CASE("Unit.Parallel.Main", "[Unit][Parallel]") {
+  const auto& cache = *(*global_cache_proxy).ckLocalBranch();
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_Main.cpp]: `global_cache_proxy->`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Parallel/Main.hpp (hunk 2 above)
Comment (kidder) [reply src/Parallel/Main.hpp:18]: clang-format orders upper-case before lower-case

Comment (nilsdeppe) [reply src/Parallel/Main.hpp:18]: Okay, that's what I figured, thanks for checking!

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**